engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraints.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.metadata.core;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.List;
+
+import javax.validation.valueextraction.ValidateUnwrappedValue;
+
+import org.hibernate.validator.internal.engine.cascading.ValueExtractorDescriptor;
+import org.hibernate.validator.internal.engine.cascading.ValueExtractorManager;
+import org.hibernate.validator.internal.metadata.descriptor.ConstraintDescriptorImpl;
+import org.hibernate.validator.internal.metadata.location.ConstraintLocation;
+import org.hibernate.validator.internal.metadata.location.TypeArgumentConstraintLocation;
+import org.hibernate.validator.internal.util.TypeHelper;
+import org.hibernate.validator.internal.util.TypeResolutionHelper;
+import org.hibernate.validator.internal.util.logging.Log;
+import org.hibernate.validator.internal.util.logging.LoggerFactory;
+
+import com.fasterxml.classmate.ResolvedType;
+
+/**
+ * Helper used to create {@link MetaConstraint}s.
+ *
+ * @author Guillaume Smet
+ */
+public class MetaConstraints {
+
+	private static final Log LOG = LoggerFactory.make();
+
+	private MetaConstraints() {
+	}
+
+	public static <A extends Annotation> MetaConstraint<A> create(TypeResolutionHelper typeResolutionHelper, ValueExtractorManager valueExtractors,
+			ConstraintDescriptorImpl<A> constraintDescriptor, ConstraintLocation location) {
+		ValueExtractorDescriptor valueExtractorDescriptor;
+		Type typeOfValidatedElement;
+
+		if ( location instanceof TypeArgumentConstraintLocation ) {
+			TypeArgumentConstraintLocation typeArgumentConstraintLocation = (TypeArgumentConstraintLocation) location;
+
+			Class<?> declaredType = TypeHelper.getErasedReferenceType( typeArgumentConstraintLocation.getDelegate().getTypeForValidatorResolution() );
+			TypeVariable<?> typeParameter = typeArgumentConstraintLocation.getTypeParameter();
+
+			valueExtractorDescriptor = valueExtractors.getCascadedValueExtractor( declaredType, typeParameter );
```

Comment:
We should rename `getCascadedValueExtractor` into `getValueExtractor` as it's not used for cascading only anymore.